internal/controller/gateway.go

```diff
@@ -31,46 +31,103 @@ import (
 )
 
 type gatewayReconciler struct {
+	ctx          context.Context
 	client       client.Client
 	eventHandler cache.ResourceEventHandler
-	logrus.FieldLogger
+	log          logrus.FieldLogger
+
+	// gatewayClassControllerName is the configured controller of managed gatewayclasses.
+	gatewayClassControllerName string
 }
 
 // NewGatewayController creates the gateway controller from mgr. The controller will be pre-configured
-// to watch for Gateway objects across all namespaces.
-func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+// to watch for Gateway objects across all namespaces and reconcile those that match class.
+func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger, gatewayClassControllerName string) (controller.Controller, error) {
 	r := &gatewayReconciler{
-		client:       mgr.GetClient(),
-		eventHandler: eventHandler,
-		FieldLogger:  log,
+		ctx:                        context.Background(),
+		client:                     mgr.GetClient(),
+		eventHandler:               eventHandler,
+		log:                        log,
+		gatewayClassControllerName: gatewayClassControllerName,
 	}
 	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return nil, err
 	}
-	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, r.enqueueRequestForOwnedGateway()); err != nil {
 		return nil, err
 	}
 	return c, nil
 }
 
+// enqueueRequestForOwnedGateway returns an event handler that maps events to
+// Gateway objects that reference a GatewayClass owned by Contour.
+func (r *gatewayReconciler) enqueueRequestForOwnedGateway() handler.EventHandler {
```

Comment:
Should we also check here that we only enqueue requests for Gateways that reference a GatewayClass that is owned by Contour *and* admitted

Reply:
Yes, definitely. Otherwise, we could be reconciling more than one GatewayClass (which is not what we are intending at the moment).

Reply:
For Gateway, the problem is that you need to enqueue it to set the status on it. So I think the logic for this is to make sure that we only select a single gateway and set errors on those that may match but don't fit the criteria (meaning oldest wins?)

Reply:
https://github.com/projectcontour/contour/issues/3864